File: prg/src/perf.ml
open Printf
module S = States

let () = 
  if Array.length Sys.argv < 3 then
    begin
      printf "Use %s d s\n" Sys.argv.(0);
      exit 1
    end
  else
    ();
  let (d, s) = (int_of_string Sys.argv.(1), int_of_string Sys.argv.(2)) in
  let space = S.state_space d s in

  let t1 = Sys.time() in

  let h = Hashtbl.create space in
  let w_set = S.compute_w d s in

  for k=0 to space-1 do
    Hashtbl.add h w_set.(k) k;
  done;

  let t2 = Sys.time() in

  let a = ref 56 in

  for k=0 to space-1 do
    let key = Hashtbl.find h w_set.(k) in
    if key > !a then
      a := key
  done;

  let t3 = Sys.time() in

  printf "Time to compute w_set and the hashtable: %f\n" (t2-.t1);
  printf "Time to access all the elements of the hashtable %f\n" (t3-.t2);
  ()

File: prg/src/c++/perf.cpp
#include <cstdio>
#include <ctime>
#include <unordered_map>
#include "states.hpp"

int main(int argc, char *argv[]){
  unsigned int d, s, space;
  clock_t t1, t2, t3;

  if (argc < 3){
    printf("Use %s d s\n", argv[0]);
    exit(1);
  }

  d = atoi(argv[1]);
  s = atoi(argv[2]);
  space = state_space(d, s);

  t1 = clock();

  std::unordered_map<W, unsigned int> h(space);
  std::vector<W> w_set = compute_w(d, s);

  for (unsigned int k = 0; k < space; k++)
    h[w_set[k]] = k;

  t2 = clock();

  unsigned int a = 56;

  for (unsigned int k = 0; k < space; k++){
    unsigned int key = h[w_set[k]];
    if (key > a)
      a = key;
  }

  t3 = clock();

  printf("Time to compute w_set and the hashtable: %f\n", ((double)(t2-t1))/CLOCKS_PER_SEC);
  printf("Time to access all the elements of the hashtable %f\n", (double)(t3-t2)/CLOCKS_PER_SEC);
	
  return 0;
}
